api/client/alpn.go

```diff
@@ -52,42 +75,105 @@ func NewALPNDialer(cfg ALPNDialerConfig) ContextDialer {
+		tlsConfig.RootCAs, err = d.cfg.GetClusterCAs(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	if d.shouldUpdateServerName() {
+		tlsConfig.ServerName, _, err = webclient.ParseHostPort(addr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return tlsConfig, nil
+}
 
-	dialer := NewDialer(ctx, d.cfg.DialTimeout, d.cfg.DialTimeout, WithTLSConfig(d.cfg.TLSConfig))
-	if d.cfg.ALPNConnUpgradeRequired {
-		dialer = newALPNConnUpgradeDialer(dialer, &tls.Config{
-			InsecureSkipVerify: d.cfg.TLSConfig.InsecureSkipVerify,
-		})
+// DialContext implements ContextDialer.
+func (d *ALPNDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	tlsConfig, err := d.getTLSConfig(ctx, addr)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
 
+	dialer := NewDialer(ctx, d.cfg.DialTimeout, d.cfg.DialTimeout,
+		WithInsecureSkipVerify(d.cfg.TLSConfig.InsecureSkipVerify),
+		WithALPNConnUpgrade(d.cfg.ALPNConnUpgradeRequired),
+		WithALPNConnUpgradePing(shouldALPNConnUpgradeWithPing(tlsConfig)),
+	)
+
 	conn, err := dialer.DialContext(ctx, network, addr)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	tlsConn := tls.Client(conn, d.cfg.TLSConfig)
+	tlsConn := tls.Client(conn, tlsConfig)
 	if err := tlsConn.HandshakeContext(ctx); err != nil {
 		defer tlsConn.Close()
 		return nil, trace.Wrap(err)
 	}
+
+	if IsALPNPingProtocol(tlsConn.ConnectionState().NegotiatedProtocol) {
+		logrus.Debugf("Using ping connection for protocol %v.", tlsConn.ConnectionState().NegotiatedProtocol)
+		return pingconn.NewTLS(tlsConn), nil
+	}
 	return tlsConn, nil
 }
 
-// DialALPN a helper to dial using an ALPNDialer and returns a tls.Conn if
-// successful.
-func DialALPN(ctx context.Context, addr string, cfg ALPNDialerConfig) (*tls.Conn, error) {
+// DialALPN a helper to dial using an ALPNDialer.
+func DialALPN(ctx context.Context, addr string, cfg ALPNDialerConfig) (net.Conn, error) {
 	conn, err := NewALPNDialer(cfg).DialContext(ctx, "tcp", addr)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	tlsConn, ok := conn.(*tls.Conn)
-	if !ok {
-		return nil, trace.BadParameter("failed to convert to tls.Conn")
+	return conn, trace.Wrap(err)
+}
+
+// IsALPNPingProtocol checks if the provided protocol is suffixed with Ping.
+func IsALPNPingProtocol(protocol string) bool {
+	return strings.HasSuffix(protocol, constants.ALPNSNIProtocolPingSuffix)
+}
+
+// shouldALPNConnUpgradeWithPing returns true if Ping wrapper is required
+// during connection upgrade.
+func shouldALPNConnUpgradeWithPing(config *tls.Config) bool {
+	for _, proto := range config.NextProtos {
+		switch proto {
+		// Server usually sends SSH keepalives or HTTP2 pings every five
+		// minutes for reverse tunnel and SSH connections. Load balancers
+		// usually have a shorter idle timeout. Thus wrapping the connection
+		// with Ping protocol at the connection upgrade layer to keepalive.
+		case constants.ALPNSNIProtocolReverseTunnel,
```

Comment:
Can we add the "teleport-reversetunnelv2" - [ProtocolReverseTunnelV2](https://github.com/gravitational/teleport/blob/922fc8bfb4046f9617bed2ed007e57a53a7c40ca/lib/srv/alpnproxy/common/protocols.go#L77) to the list. 


Also what about handling `teleport-auth@` protocol that is used in TLS Routing to dial auth service through teleport proxy but under the hoop HTTP2/grpc protocol is used. 

Note: that [ConfigureALP](https://github.com/gravitational/teleport/blob/ff3c8699a830eafe364180f6ba80c711533c4e32/api/client/client.go#L455) where `teleport-auth@` extension is added after the `shouldALPNConnUpgradeWithPing` function is called.

Reply:
by definition `teleport-reversetunnelv2` is always used with `teleport-reversetunnel` so I don't think it's needed here.

Added ping in `authConnect`